File: bedrock/src/worldgen/logic/layouts.test.ts
import { describe, expect, it } from 'vitest';
import { generatedFiles } from './generate';
import { BLOCK, type Layout, NPC_ENTITY, type Vec3, allLayouts, facingYaw, signPixels, structureId, villageStyleForBiome } from './layouts';
import { BLOCK_DATA_VERSION, buildPalette, encodeMcstructure, structurePath } from './mcstructure';
import { TAG, type Tag, child, readNbt } from './nbt';

// node:fs without pulling @types/node into the (Bedrock) type program
interface Fs {
  readFileSync(p: string, enc?: string): Uint8Array & { toString(): string };
  existsSync(p: string): boolean;
  readdirSync(p: string): string[];
}
const fs = (await import(/* @vite-ignore */ ['node', 'fs'].join(':'))) as Fs;
const PACK = 'packs/worldgen';

function langKeys(): Set<string> {
  const keys = new Set<string>();
  for (const dir of fs.readdirSync('lang')) {
    const f = `lang/${dir}/en_US.lang`;
    if (!fs.existsSync(f)) continue;
    for (const line of String(fs.readFileSync(f, 'utf8')).split('\n')) {
      const m = /^([^#=][^=]*)=/.exec(line);
      if (m) keys.add(m[1] as string);
    }
  }
  return keys;
}

const layouts = [...allLayouts().values()];
const byId = (id: string): Layout => allLayouts().get(id) as Layout;

function count<T>(items: readonly T[], pred: (t: T) => boolean): number {
  return items.filter(pred).length;
}

describe('casino layouts (GAME_DESIGN §16)', () => {
  it('has one village palette per village type plus the Parlor and the Lounge', () => {
    expect(layouts.map((l) => l.id).sort()).toEqual(
      [
        'high_roller_lounge',
        'piglin_parlor',
        'village_casino_desert',
        'village_casino_plains',
        'village_casino_savanna',
        'village_casino_snowy',
        'village_casino_taiga',
      ].sort(),
    );
  });

  it('uses the spec bounding boxes', () => {
    for (const l of layouts.filter((l) => l.kind === 'village_casino')) expect(l.size).toEqual({ x: 17, y: 10, z: 17 });
    expect(byId('piglin_parlor').size).toEqual({ x: 21, y: 12, z: 21 });
    expect(byId('high_roller_lounge').size).toEqual({ x: 15, y: 9, z: 15 });
  });

  it('village casino contents: cashier, blackjack, roulette, 3 copper, 1 gold, wheel, loan shark, croupier, 1 chest', () => {
    for (const l of layouts.filter((l) => l.kind === 'village_casino')) {
      const blocks = l.tables.map((t) => t.block);
      expect(count(blocks, (b) => b === BLOCK.cashier)).toBe(1);
      expect(count(blocks, (b) => b === BLOCK.blackjack)).toBe(1);
      expect(count(blocks, (b) => b === BLOCK.roulette)).toBe(1);
      expect(count(blocks, (b) => b === BLOCK.slotsCopper)).toBe(3);
      expect(count(blocks, (b) => b === BLOCK.slotsGold)).toBe(1);
      expect(count(blocks, (b) => b === BLOCK.wheel)).toBe(1);
      expect(count(blocks, (b) => b === BLOCK.uth)).toBe(1);
      expect(l.npcs.map((n) => n.role).sort()).toEqual(['croupier', 'loan_shark']);
      expect(l.chests).toEqual([expect.objectContaining({ loot: 'village_casino' })]);
    }
  });

  it('Piglin Parlor contents: craps, low-stakes poker, baccarat, 2 gold slots, plinko, nether cashier, 2 dealers, moneylender', () => {
    const l = byId('piglin_parlor');
    const blocks = l.tables.map((t) => t.block);
    expect(count(blocks, (b) => b === BLOCK.craps)).toBe(1);
    expect(l.tables.filter((t) => t.block === BLOCK.poker).map((t) => t.preset)).toEqual(['parlor_poker']);
    expect(count(blocks, (b) => b === BLOCK.slotsGold)).toBe(2);
    expect(count(blocks, (b) => b === BLOCK.plinko)).toBe(1);
    expect(count(blocks, (b) => b === BLOCK.netherCashier)).toBe(1);
    expect(count(blocks, (b) => b === BLOCK.baccarat)).toBe(1);
    expect(l.npcs.map((n) => n.role).sort()).toEqual(['piglin_dealer', 'piglin_dealer', 'piglin_moneylender']);
    expect(l.chests.map((c) => c.loot)).toEqual(['piglin_parlor']);
  });

  it('High Roller Lounge contents: 2 netherite slots, HR blackjack + roulette + baccarat + UTH, cashier, shulker croupier, baccarat dealer', () => {
    const l = byId('high_roller_lounge');
    const blocks = l.tables.map((t) => t.block);
    expect(count(blocks, (b) => b === BLOCK.slotsNetherite)).toBe(2);
    expect(l.tables.filter((t) => t.block === BLOCK.blackjackHighRoller).map((t) => t.preset)).toEqual(['high_roller_blackjack']);
    expect(l.tables.filter((t) => t.block === BLOCK.rouletteHighRoller).map((t) => t.preset)).toEqual(['high_roller_roulette']);
    expect(count(blocks, (b) => b === BLOCK.cashier)).toBe(1);
    expect(l.tables.filter((t) => t.block === BLOCK.baccaratHighRoller).map((t) => t.preset)).toEqual(['high_roller_baccarat']);
    expect(l.tables.filter((t) => t.block === BLOCK.uthHighRoller).map((t) => t.preset)).toEqual(['high_roller_uth']);
    expect(l.size).toEqual({ x: 15, y: 9, z: 15 });
    expect(l.npcs.map((n) => n.role)).toEqual(['shulker_croupier', 'baccarat_dealer']);
    expect(l.chests.map((c) => c.loot)).toEqual(['high_roller']);
  });

  it('references table blocks and NPC entities by their STRINGS.md ids', () => {
    const keys = langKeys();
    for (const id of Object.values(BLOCK)) expect(keys.has(`block.burmaldaholic.${id.slice('burmaldaholic:'.length)}`), id).toBe(true);
    for (const ids of Object.values(NPC_ENTITY))
      for (const id of ids) expect(keys.has(`entity.burmaldaholic.${id.slice('burmaldaholic:'.length)}`), id).toBe(true);
  });

  const walkable = (l: Layout, occupied: Set<string>, x: number, z: number): boolean => {
    if (x < 0 || z < 0 || x >= l.size.x || z >= l.size.z) return false;
    if (occupied.has(`${x},${z}`)) return false;
    const feet = l.grid.get(x, 1, z)?.name;
    const head = l.grid.get(x, 2, z)?.name;
    return (feet === 'minecraft:air' || !!feet?.endsWith('_carpet')) && head === 'minecraft:air';
  };

  for (const l of layouts) {
    it(`${l.id}: furniture sits inside, on free floor cells, and is reachable from the door`, () => {
      const occupied = new Set<string>();
      const check = (p: Vec3, what: string) => {
        expect(p.y, what).toBe(1);
        expect(p.x > 0 && p.z > 0 && p.x < l.size.x - 1 && p.z < l.size.z - 1, `${what} inside walls`).toBe(true);
        const floor = l.grid.get(p.x, 0, p.z)?.name;
        expect(floor && floor !== 'minecraft:air', `${what} has a floor`).toBe(true);
        const key = `${p.x},${p.z}`;
        expect(occupied.has(key), `${what} overlaps`).toBe(false);
        occupied.add(key);
      };
      for (const t of l.tables) {
        check(t.pos, t.block);
        const cell = l.grid.get(t.pos.x, 1, t.pos.z)?.name ?? '';
        expect(cell === 'minecraft:air' || cell.endsWith('_carpet'), `${t.block} replaces air/carpet, not ${cell}`).toBe(true);
      }
      for (const n of l.npcs) {
        check(n.pos, n.role);
        expect(l.grid.get(n.pos.x, 2, n.pos.z)?.name, `${n.role} headroom`).toBe('minecraft:air');
      }
      for (const c of l.chests) {
        check(c.pos, 'chest');
        expect(l.grid.get(c.pos.x, 1, c.pos.z)?.name).toBe('minecraft:chest');
      }
      // flood fill from the door
      const start: [number, number][] = [];
      for (let x = l.door.x0; x <= l.door.x1; x++) {
        expect(l.grid.get(x, 1, l.size.z - 1)?.name, 'door is open').toBe('minecraft:air');
        start.push([x, l.size.z - 1]);
      }
      const seen = new Set(start.map(([x, z]) => `${x},${z}`));
      const queue = [...start];
      while (queue.length) {
        const [x, z] = queue.shift() as [number, number];
        for (const [dx, dz] of [
          [1, 0],
          [-1, 0],
          [0, 1],
          [0, -1],
        ] as const) {
          const k = `${x + dx},${z + dz}`;
          if (!seen.has(k) && walkable(l, occupied, x + dx, z + dz)) {
            seen.add(k);
            queue.push([x + dx, z + dz]);
          }
        }
      }
      const reachable = (p: Vec3) =>
        [
          [1, 0],
          [-1, 0],
          [0, 1],
          [0, -1],
        ].some(([dx, dz]) => seen.has(`${p.x + (dx as number)},${p.z + (dz as number)}`));
      for (const t of l.tables) expect(reachable(t.pos), `${t.block} at ${t.pos.x},${t.pos.z} reachable`).toBe(true);
      for (const n of l.npcs) expect(reachable(n.pos), `${n.role} reachable`).toBe(true);
      for (const c of l.chests) expect(reachable(c.pos), 'chest reachable').toBe(true);
    });
  }

  it('spells CASINO on the village sign (3×5 font, 16 columns)', () => {
    const s = signPixels('CASINO');
    expect(s.width).toBe(16);
    expect(new Set(s.pixels.map((p) => p.letter))).toEqual(new Set([0, 1, 2, 3, 4, 5]));
    expect(s.pixels.every((p) => p.row >= 0 && p.row < 5 && p.col >= 0 && p.col < 16)).toBe(true);
    const v = byId('village_casino_plains');
    const lit = (x: number, y: number) => ['minecraft:glowstone', 'minecraft:lit_redstone_lamp'].includes(v.grid.get(x, y, 16)?.name ?? '');
    for (const p of s.pixels) expect(lit(p.col, 9 - p.row)).toBe(true);
    // every lamp is powered from behind
    for (let x = 0; x < 17; x++)
      for (let y = 5; y < 10; y++) if (v.grid.get(x, y, 16)?.name === 'minecraft:lit_redstone_lamp') expect(v.grid.get(x, y, 15)?.name).toBe('minecraft:redstone_block');
  });

  it('maps biomes to village styles', () => {
    expect(villageStyleForBiome('minecraft:desert')).toBe('desert');
    expect(villageStyleForBiome('minecraft:savanna_plateau')).toBe('savanna');
    expect(villageStyleForBiome('minecraft:taiga')).toBe('taiga');
    expect(villageStyleForBiome('minecraft:snowy_plains')).toBe('snowy');
    expect(villageStyleForBiome('minecraft:ice_plains')).toBe('snowy');
    expect(villageStyleForBiome('minecraft:plains')).toBe('plains');
    expect(villageStyleForBiome('minecraft:meadow')).toBe('plains');
  });

  it('yaw faces the given direction', () => {
    expect(facingYaw('south')).toBe(0);
    expect(facingYaw('north')).toBe(180);
    expect(facingYaw('west')).toBe(90);
    expect(facingYaw('east')).toBe(-90);
    expect(structureId('piglin_parlor')).toBe('burmaldaholic:worldgen_piglin_parlor');
  });
});

describe('.mcstructure templates', () => {
  const tagList = (t: Tag | undefined): Tag[] => (t?.type === TAG.list ? t.value : []);
  const ints = (t: Tag | undefined): number[] => tagList(t).map((x) => (x.type === TAG.int ? x.value : NaN));

  for (const l of layouts) {
    it(`${l.id}: valid little-endian NBT with the Bedrock structure schema`, () => {
      const { name, tag } = readNbt(encodeMcstructure(l.grid));
      expect(name).toBe('');
      expect(child(tag, 'format_version')).toEqual({ type: TAG.int, value: 1 });
      expect(ints(child(tag, 'size'))).toEqual([l.size.x, l.size.y, l.size.z]);
      expect(ints(child(tag, 'structure_world_origin'))).toEqual([0, 0, 0]);
      const structure = child(tag, 'structure') as Tag;
      const layers = tagList(child(structure, 'block_indices'));
      expect(layers).toHaveLength(2);
      const n = l.size.x * l.size.y * l.size.z;
      const primary = ints(layers[0]);
      expect(primary).toHaveLength(n);
      expect(ints(layers[1])).toEqual(new Array(n).fill(-1));
      expect(tagList(child(structure, 'entities'))).toEqual([]);
      const palette = tagList(child(child(child(structure, 'palette') as Tag, 'default') as Tag, 'block_palette'));
      expect(palette.length).toBe(buildPalette(l.grid).palette.length);
      for (const p of palette) {
        const nm = child(p, 'name');
        expect(nm?.type === TAG.string && nm.value.startsWith('minecraft:')).toBe(true);
        expect(child(p, 'version')).toEqual({ type: TAG.int, value: BLOCK_DATA_VERSION });
        expect(child(p, 'states')?.type).toBe(TAG.compound);
      }
      expect(Math.max(...primary)).toBe(palette.length - 1);
      expect(Math.min(...primary)).toBeGreaterThanOrEqual(-1);
      // never contains another module's blocks: those are placed by script
      expect(palette.some((p) => JSON.stringify(child(p, 'name')).includes('burmaldaholic'))).toBe(false);
    });
  }

  it('committed templates match the generator (run: node src/worldgen/tools/gen-assets.mjs)', () => {
    for (const f of generatedFiles()) {
      const file = `${PACK}/${f.path}`;
      expect(fs.existsSync(file), file).toBe(true);
      expect(bytesEqual(fs.readFileSync(file), f.bytes), `${file} is stale`).toBe(true);
    }
    expect(generatedFiles().map((f) => f.path)).toEqual(layouts.map(structurePath));
  });
});

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

File: java/src/test/java/dev/nezo/burmaldaholic/worldgen/logic/LayoutsTest.java
package dev.nezo.burmaldaholic.worldgen.logic;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.Test;

class LayoutsTest {
	private static long count(Layout l, String block) {
		long n = 0;
		Vec s = l.size();
		for (int x = 0; x < s.x(); x++) {
			for (int y = 0; y < s.y(); y++) {
				for (int z = 0; z < s.z(); z++) {
					BlockSpec b = l.grid().get(x, y, z);
					if (b != null && b.is(block)) {
						n++;
					}
				}
			}
		}
		return n;
	}

	private static long npcs(Layout l, NpcRole role) {
		return l.markers().stream().filter(m -> m.marker() instanceof Markers.Npc n && n.role() == role).count();
	}

	@Test
	void sevenLayoutsWithSpecSizes() {
		assertEquals(7, Layouts.all().size());
		for (VillageStyle s : VillageStyle.values()) {
			assertEquals(new Vec(17, 10, 17), Layouts.village(s).size(), s.id());
			assertEquals(CasinoKind.VILLAGE_CASINO, Layouts.village(s).kind());
		}
		assertEquals(new Vec(21, 12, 21), Layouts.piglinParlorLayout().size());
		assertEquals(new Vec(15, 9, 15), Layouts.highRollerLayout().size());
	}

	@Test
	void villageContentsPerSpec() {
		for (VillageStyle s : VillageStyle.values()) {
			Layout l = Layouts.village(s);
			assertEquals(1, count(l, Layouts.CASHIER));
			assertEquals(1, count(l, Layouts.BLACKJACK));
			assertEquals(1, count(l, Layouts.ROULETTE));
			assertEquals(3, count(l, Layouts.SLOTS_COPPER));
			assertEquals(1, count(l, Layouts.SLOTS_GOLD));
			assertEquals(1, count(l, Layouts.WHEEL));
			assertEquals(1, count(l, Layouts.UTH));
			assertEquals(1, npcs(l, NpcRole.LOAN_SHARK));
			assertEquals(1, npcs(l, NpcRole.CROUPIER));
			assertEquals(1, l.markers().stream().filter(m -> m.marker() instanceof Markers.Chest c
				&& c.table() == CasinoLoot.Table.VILLAGE_CASINO).count());
			assertEquals(1, l.jigsaws().size(), "street connector");
			assertTrue(count(l, "minecraft:glowstone") > 0 && count(l, "minecraft:redstone_lamp") > 0, "CASINO sign");
		}
	}

	@Test
	void villagePalettesDiffer() {
		Set<String> floors = new HashSet<>();
		for (VillageStyle s : VillageStyle.values()) {
			floors.add(Layouts.village(s).grid().get(8, 0, 8).name() + Layouts.village(s).grid().get(0, 2, 0).name());
		}
		assertEquals(VillageStyle.values().length, floors.size(), "each village type has its own palette");
	}

	@Test
	void parlorContentsPerSpec() {
		Layout l = Layouts.piglinParlorLayout();
		assertEquals(1, count(l, Layouts.CRAPS));
		assertEquals(1, count(l, Layouts.POKER));
		assertEquals(2, count(l, Layouts.SLOTS_GOLD));
		assertEquals(1, count(l, Layouts.PLINKO));
		assertEquals(1, count(l, Layouts.NETHER_CASHIER));
		assertEquals(1, count(l, Layouts.BACCARAT));
		assertEquals(2, npcs(l, NpcRole.PIGLIN_DEALER));
		assertEquals(1, npcs(l, NpcRole.PIGLIN_MONEYLENDER));
		// rows 9..11 keep the Nether rock (structure void)
		assertNull(l.grid().get(10, 10, 10));
		// back door towards the bastion + main entrance
		assertTrue(l.grid().get(10, 2, 0).is("minecraft:air") && l.grid().get(10, 2, 20).is("minecraft:air"));
	}

	@Test
	void loungeContentsPerSpec() {
		Layout l = Layouts.highRollerLayout();
		assertEquals(2, count(l, Layouts.SLOTS_NETHERITE));
		assertEquals(1, count(l, Layouts.BLACKJACK_HIGH_ROLLER));
		assertEquals(1, count(l, Layouts.ROULETTE_HIGH_ROLLER));
		assertEquals(1, count(l, Layouts.CASHIER));
		assertEquals(1, count(l, Layouts.BACCARAT_HIGH_ROLLER));
		assertEquals(1, count(l, Layouts.UTH_HIGH_ROLLER));
		assertEquals(1, npcs(l, NpcRole.SHULKER_CROUPIER));
		assertEquals(1, npcs(l, NpcRole.BACCARAT_DEALER));
	}

	@Test
	void markersSitInClearCellsInsideAndEveryLayoutHasOneAnchor() {
		for (Layout l : Layouts.all().values()) {
			long anchors = l.markers().stream().filter(m -> m.marker() instanceof Markers.Anchor a && a.layoutId().equals(l.id())).count();
			assertEquals(1, anchors, l.id());
			Set<Vec> seen = new HashSet<>();
			for (Layout.PlacedMarker m : l.markers()) {
				Vec p = m.pos();
				assertTrue(l.grid().inside(p.x(), p.y(), p.z()), l.id() + " marker outside " + p);
				assertTrue(seen.add(p), l.id() + " two markers at " + p);
				if (m.marker() instanceof Markers.Npc) {
					// NPC stands in the cell below the marker: must not be a solid wall/table
					BlockSpec stand = l.grid().get(p.x(), p.y() - 1, p.z());
					assertNotNull(stand);
					assertTrue(stand.is("minecraft:air") || stand.name().endsWith("_carpet"), l.id() + " npc stands in " + stand);
					BlockSpec head = l.grid().get(p.x(), p.y() + 1, p.z());
					assertTrue(head.is("minecraft:air"), l.id() + " npc head in " + head);
				}
			}
		}
	}

	@Test
	void tablesFaceSomewhereAndAreOnTheFloorRow() {
		for (Layout l : Layouts.all().values()) {
			Vec s = l.size();
			for (int x = 0; x < s.x(); x++) {
				for (int y = 0; y < s.y(); y++) {
					for (int z = 0; z < s.z(); z++) {
						BlockSpec b = l.grid().get(x, y, z);
						if (b != null && b.name().startsWith("burmaldaholic:")) {
							assertEquals(1, y, l.id() + " table off the floor row: " + b);
							assertTrue(b.properties().containsKey("facing"), b.toString());
						}
					}
				}
			}
		}
	}

	@Test
	void entranceIsOpenOnPlusZ() {
		for (Layout l : Layouts.all().values()) {
			int midX = l.size().x() / 2;
			int z = l.size().z() - 1;
			assertTrue(l.grid().get(midX, 1, z).is("minecraft:air") && l.grid().get(midX, 2, z).is("minecraft:air"), l.id());
		}
	}

	@Test
	void signSpellsCasino() {
		Layouts.Sign sign = Layouts.signPixels("CASINO");
		assertEquals(3 + 3 + 3 + 1 + 3 + 3, sign.width());
		List<Integer> letters = new ArrayList<>();
		sign.pixels().forEach(p -> {
			if (!letters.contains(p.letter())) {
				letters.add(p.letter());
			}
		});
		assertEquals(List.of(0, 1, 2, 3, 4, 5), letters);
		assertFalse(sign.pixels().isEmpty());
	}

	@Test
	void templateNbtHasPaletteBlocksAndMarkers() {
		for (Layout l : Layouts.all().values()) {
			Map<String, Object> root = TemplateExporter.template(l);
			assertEquals(TemplateExporter.DATA_VERSION, root.get("DataVersion"));
			Nbt.TagList blocks = (Nbt.TagList) root.get("blocks");
			Nbt.TagList palette = (Nbt.TagList) root.get("palette");
			long structureBlocks = blocks.items().stream().filter(b -> {
				@SuppressWarnings("unchecked") Map<String, Object> m = (Map<String, Object>) b;
				@SuppressWarnings("unchecked") Map<String, Object> state = (Map<String, Object>) palette.items().get((Integer) m.get("state"));
				return state.get("Name").equals("minecraft:structure_block");
			}).count();
			assertEquals(l.markers().size(), structureBlocks, l.id());
			for (Object b : blocks.items()) {
				@SuppressWarnings("unchecked") Map<String, Object> m = (Map<String, Object>) b;
				assertTrue((Integer) m.get("state") < palette.items().size());
			}
		}
	}

	@Test
	void villageStyleMapping() {
		assertEquals(VillageStyle.SNOWY, VillageStyle.forBiome("minecraft:snowy_plains"));
		assertEquals(VillageStyle.DESERT, VillageStyle.forBiome("minecraft:desert"));
		assertEquals(VillageStyle.SAVANNA, VillageStyle.forBiome("minecraft:savanna_plateau"));
		assertEquals(VillageStyle.TAIGA, VillageStyle.forBiome("minecraft:old_growth_spruce_taiga"));
		assertEquals(VillageStyle.PLAINS, VillageStyle.forBiome("minecraft:meadow"));
		assertEquals(VillageStyle.PLAINS, VillageStyle.byCasinoPool("minecraft:village/plains/terminators").orElseThrow());
		assertTrue(VillageStyle.byCasinoPool("minecraft:village/plains/zombie/terminators").isEmpty(), "no casinos in zombie villages");
		assertTrue(VillageStyle.byCasinoPool("minecraft:village/plains/houses").isEmpty());
	}
}
